src/engine/renderer/glsl_source/computeLight_fp.glsl

```diff
@@ -32,196 +32,199 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
-  // > glTF stores occlusion in the red (R) channel, allowing it to optionally share the same image
-  // > with the roughness and metallic channels.
-  float roughness = materialColor.g;
-  float metalness = materialColor.b;
+	vec3 H = normalize( lightDir + viewDir );
+
+	#if defined(USE_PHYSICAL_MAPPING) || defined(r_specularMapping)
+		float NdotH = clamp( dot( normal, H ), 0.0, 1.0 );
+	#endif // USE_PHYSICAL_MAPPING || r_specularMapping
+
+	// clamp( NdotL, 0.0, 1.0 ) is done below
+	float NdotL = dot( normal, lightDir );
+
+	#if !defined(USE_BSP_SURFACE) && defined(r_halfLambertLighting)
+		// http://developer.valvesoftware.com/wiki/Half_Lambert
+		NdotL = NdotL * 0.5 + 0.5;
+		NdotL *= NdotL;
+	#endif
+
+	NdotL = clamp( NdotL, 0.0, 1.0 );
+
+	#if defined(USE_PHYSICAL_MAPPING)
+		// Daemon PBR packing defaults to ORM like glTF 2.0 defines
+		// https://www.khronos.org/blog/art-pipeline-for-gltf
+		// > ORM texture for Occlusion, Roughness, and Metallic
+		// https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/schema/material.pbrMetallicRoughness.schema.json
+		// > The metalness values are sampled from the B channel. The roughness values are sampled from the G channel.
+		// > These values are linear. If other channels are present (R or A), they are ignored for metallic-roughness calculations.
+		// https://docs.blender.org/manual/en/2.80/addons/io_scene_gltf2.html
+		// > glTF stores occlusion in the red (R) channel, allowing it to optionally share the same image
+		// > with the roughness and metallic channels.
+		float roughness = materialColor.g;
+		float metalness = materialColor.b;
+
+		float NdotV = clamp( dot( normal, viewDir ), 0.0, 1.0);
+		float VdotH = clamp( dot( viewDir, H ), 0.0, 1.0);
+
+		float alpha = roughness * roughness;
+		float k = 0.125 * ( roughness + 1.0 ) * ( roughness + 1.0 );
+
+		float D = alpha / ( ( NdotH * NdotH ) * (alpha * alpha - 1.0 ) + 1.0 );
+		D *= D;
+
+		float FexpNH = pow( 1.0 - NdotH, 5.0 );
+		float FexpNV = pow( 1.0 - NdotV, 5.0 );
+		vec3 F = mix( vec3( 0.04 ), diffuseColor.rgb, metalness );
+		F += ( 1.0 - F ) * FexpNH;
+
+		float G = NdotL / (NdotL * ( 1.0 - k ) + k );
+		G *= NdotV / ( NdotV * ( 1.0 - k ) + k );
+
+		vec3 diffuseBRDF = NdotL * diffuseColor.rgb * ( 1.0 - metalness );
+		vec3 specularBRDF = vec3( ( D * F * G ) / max( 4.0 * NdotL * NdotV, 0.0001f ) );
+		color.rgb += ( diffuseBRDF + specularBRDF ) * lightColor.rgb * NdotL;
+		color.a = mix( diffuseColor.a, 1.0, FexpNV );
+	#else // !USE_PHYSICAL_MAPPING
+
+	#if defined(USE_REFLECTIVE_SPECULAR)
+		// not implemented for PBR yet
+		vec4 envColor0 = textureCube(u_EnvironmentMap0, reflect( -viewDir, normal ) );
+		vec4 envColor1 = textureCube(u_EnvironmentMap1, reflect( -viewDir, normal ) );
+
+		materialColor.rgb *= mix( envColor0, envColor1, u_EnvironmentInterpolation ).rgb;
+	#endif // USE_REFLECTIVE_SPECULAR
+
+	color.rgb += lightColor.rgb * NdotL * diffuseColor.rgb;
```

Comment:
This block should be moved one tab right.

Reply:
Done.